### PR TITLE
sysupdate: Allow patterns to match path with directories

### PR DESCRIPTION
**Pull request by valentindavid**
`MatchPattern` for regular-file and directory as target can now match subdirectories This is useful to install files for examples in `.extra.d` directories:

```
[Target]
Type=regular-file
Path=/EFI/Linux
PathRelativeTo=boot
MatchPattern=gnomeos_@v.efi.extra.d/apparmor.addon.efi
```

The if the directories in the path do not exist, they will be created.  Whereas the part in `Path` is not created.

### CODE REVIEW

File: src/sysupdate/sysupdate-pattern.c [hunk 1]
```diff
@@ -203,6 +223,17 @@ int pattern_match(const char *pattern, const char *s, InstanceMetadata *ret) {
                 _cleanup_free_ char *t = NULL;
                 const char *n;
 
+                if (e->type == PATTERN_SLASH) {
+                        if (*p == '/') {
+                                ++p;
+                                continue;
+                        } else if (*p == '\0') {
```
Comment (bluca) [inline src/sysupdate/sysupdate-pattern.c]: no brackets around single-line ifs

Comment (valentindavid) [reply src/sysupdate/sysupdate-pattern.c]: Sorry, I always forget this one.

**Comment by bluca** [comment]
Documentation needs to be updated. Also please add test coverage in TEST-72-SYSUPDATE

File: src/sysupdate/sysupdate-resource.c [hunk 2]
```diff
@@ -132,16 +119,40 @@ static int resource_load_from_directory(
...
-                if ((st.st_mode & S_IFMT) != m)
+                if (!(S_ISDIR(st.st_mode) && S_ISREG(m)) && ((st.st_mode & S_IFMT) != m))
                         continue;
 
-                r = pattern_match_many(rr->patterns, de->d_name, &extracted_fields);
-                if (r < 0)
+                if (relpath != NULL)
+                        rel_joined = path_join(relpath, de->d_name);
+                else
+                        rel_joined = strdup(de->d_name);
```
Comment (bluca) [inline src/sysupdate/sysupdate-resource.c]: oom check after either call

**Review by bluca — APPROVED** [review]

File: src/sysupdate/sysupdate-resource.c [hunk 3]
```diff
@@ -111,7 +98,7 @@ static int resource_load_from_directory(
                         break;
 
                 case DT_DIR:
-                        if (m != S_IFDIR)
+                        if ((m != S_IFDIR) && (m != S_IFREG))
```
Comment (bluca) [inline src/sysupdate/sysupdate-resource.c]: inner brackets are not necessary

File: src/sysupdate/sysupdate-resource.c [hunk 4]
```diff
@@ -132,16 +119,42 @@ static int resource_load_from_directory(
                         return log_error_errno(errno, "Failed to stat %s/%s: %m", rr->path, de->d_name);
                 }
 
-                if ((st.st_mode & S_IFMT) != m)
+                if (!(S_ISDIR(st.st_mode) && S_ISREG(m)) && ((st.st_mode & S_IFMT) != m))
                         continue;
 
-                r = pattern_match_many(rr->patterns, de->d_name, &extracted_fields);
-                if (r < 0)
+                if (relpath != NULL)
```
Comment (poettering) [inline src/sysupdate/sysupdate-resource.c]: as per coding style, for pointer null checks we rely on C's downgrade to bool feature, i.e. change to `if (relpath)`

File: src/sysupdate/sysupdate-pattern.c [hunk 5]
```diff
@@ -399,13 +429,19 @@ int pattern_match(const char *pattern, const char *s, InstanceMetadata *ret) {
...
         if (ret)
                 *ret = (InstanceMetadata) INSTANCE_METADATA_NULL;
 
         return false;
+
+retry:
+        if (ret)
+                *ret = (InstanceMetadata) INSTANCE_METADATA_NULL;
+
+        return 2;
```
Comment (poettering) [inline src/sysupdate/sysupdate-pattern.c]: uh, no, please don't. let's not user arbitrary constant returns. make this an anonymous enum please.

```
enum {
        PATTERN_MATCH_NO,
        PATTERN_MATCH_YES,
        PATTERN_MATCH_RETRY,
};
```

File: src/sysupdate/sysupdate-resource.c [hunk 6]
```diff
@@ -132,16 +119,42 @@ static int resource_load_from_directory(
...
+                        rel_joined = path_join(relpath, de->d_name);
+                else
+                        rel_joined = strdup(de->d_name);
```
Comment (poettering) [inline src/sysupdate/sysupdate-resource.c]: path_join() happily skips over NULL parameters btw, hence you can just drop this conditionalization, and call path_join() in all cases

File: src/sysupdate/sysupdate-resource.c [hunk 7]
```diff
@@ -132,16 +119,42 @@ static int resource_load_from_directory(
...
+                if (r == 2) {
+                        _cleanup_closedir_ DIR *subdir = NULL;
+                        _cleanup_close_ int subfd = -EBADF;
+
+                        subfd = openat(dirfd(d), rel_joined, O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY);
+                        if (subfd < 0)
+                                continue;
+                        subdir = fdopendir(subfd);
+                        if (!subdir)
+                                continue;
```
Comment (poettering) [inline src/sysupdate/sysupdate-resource.c:134]: xopendirat()

File: src/sysupdate/sysupdate-resource.c [hunk 8]
```diff
@@ -160,6 +173,29 @@ static int resource_load_from_directory(
...
+                Resource *rr,
+                mode_t m) {
+        _cleanup_closedir_ DIR *d = NULL;
+
+        assert(rr);
+        assert(IN_SET(rr->type, RESOURCE_TAR, RESOURCE_REGULAR_FILE, RESOURCE_DIRECTORY, RESOURCE_SUBVOLUME));
+        assert(IN_SET(m, S_IFREG, S_IFDIR));
+
+        d = opendir(rr->path);
+
```
Comment (poettering) [inline src/sysupdate/sysupdate-resource.c]: we try to keep function calls and their immediate error handling together (and instead put sepating empty lines before the function call or after the error handline), hence please drop this empty line

File: src/sysupdate/sysupdate-resource.c [hunk 9]
```diff
@@ -160,6 +173,29 @@ static int resource_load_from_directory(
...
+        if (!d) {
+                if (errno == ENOENT) {
+                        log_debug("Directory %s does not exist, not loading any resources.", rr->path);
```
Comment (poettering) [inline src/sysupdate/sysupdate-resource.c]: actually, we should use `log_debug_errno(errno, ` here too. When we generate structured logging the error will be passed in the ERRNO= field, hence as a matter of coding tyle we'll always use the `log_xyz_errno()` flavour of the log calls if we have an error number, and here we do.

(i understand this is just moving around of pre-existing code, but if you touch it, please also update to newest coding style)

File: src/sysupdate/sysupdate-resource.c [hunk 10]
```diff
@@ -204,7 +240,7 @@ static int resource_load_from_blockdev(Resource *rr) {
                 r = pattern_match_many(rr->patterns, pinfo.label, &extracted_fields);
                 if (r < 0)
                         return log_error_errno(r, "Failed to match pattern: %m");
-                if (r == 0)
+                if ((r == 0) || (r == 2))
```
Comment (poettering) [inline src/sysupdate/sysupdate-resource.c]: IN_SET()

File: src/sysupdate/sysupdate-pattern.c [hunk 11]
```diff
@@ -123,7 +124,7 @@ static int pattern_split(
 
                         /* We insist that two pattern field markers are separated by some literal string that
                          * we can use to separate the fields when parsing. */
-                        if (!last_literal)
+                        if ((!last_literal) && (!last_slash))
```
Comment (poettering) [inline src/sysupdate/sysupdate-pattern.c]: the inner () are redundant btw.

File: src/sysupdate/sysupdate-resource.c [hunk 12]
```diff
@@ -69,30 +69,17 @@ static int resource_add_instance(
...
-                        return 0;
-                }
-
-                return log_error_errno(errno, "Failed to open directory '%s': %m", rr->path);
-        }
-
         for (;;) {
                 _cleanup_(instance_metadata_destroy) InstanceMetadata extracted_fields = INSTANCE_METADATA_NULL;
                 _cleanup_free_ char *joined = NULL;
+                _cleanup_free_ char *rel_joined = NULL;
```
Comment (poettering) [inline src/sysupdate/sysupdate-resource.c]: please merge with preceeding line

File: src/sysupdate/sysupdate-resource.c [hunk 13]
```diff
@@ -111,7 +98,7 @@ static int resource_load_from_directory(
                         break;
 
                 case DT_DIR:
-                        if (m != S_IFDIR)
+                        if (m != S_IFDIR && m != S_IFREG)
```
Comment (poettering) [inline src/sysupdate/sysupdate-resource.c]:
```c
if (!IN_SET(m, S_IFDIR, S_IFREG))
```

File: src/sysupdate/sysupdate-resource.c [hunk 14]
```diff
@@ -69,30 +69,17 @@ static int resource_add_instance(
         return 0;
 }
 
-static int resource_load_from_directory(
+static int resource_load_from_directory_rec(
```
Comment (poettering) [inline src/sysupdate/sysupdate-resource.c]: no needless abbreviations please, just spell out `_recursive`

File: src/sysupdate/sysupdate-resource.c [hunk 15]
```diff
@@ -132,16 +119,36 @@ static int resource_load_from_directory(
...
+                        if (r == 0)
+                                continue;
+                }
+                else if (r < 0)
                         return log_error_errno(r, "Failed to match pattern: %m");
-                if (r == 0)
+                else if (r == PATTERN_MATCH_NO)
+                        continue;
+
+                if ((de->d_type == DT_DIR) && (m != S_IFDIR))
```
Comment (poettering) [inline src/sysupdate/sysupdate-resource.c]: inner () are redundant

**Comment by poettering** [comment]
lgtm, just some superficialities left.

**Comment by poettering** [comment]
(btw, please mark all review you addressed with "Resolve conversation".)

File: src/sysupdate/sysupdate-resource.c (hunk 8 above)
Comment (poettering) [reply src/sysupdate/sysupdate-resource.c]: this hasn't been addressed btw

**Pull request merged**